File: 33.搜索旋转排序数组.js
/*
 * @lc app=leetcode.cn id=33 lang=javascript
 *
 * [33] 搜索旋转排序数组
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function (nums, target) {
    if (!nums.length) {
        return -1;
    }
    let max = nums.length - 1;
    let min = 0;
    while (max >= min) {
        let mid = Math.floor((max + min) / 2);
        // 若相等, 直接输出
        if (nums[mid] === target) {
            return mid;
        }
        // max-min相邻的情况下, mid一般直接去min, 此时需要额外判断下max的情况
        if (max === min + 1) {
            if (nums[max] === target) {
                return max;
            }
        }
        // 如果第一项直接相等, 也需要特殊判断  
        if (nums[0] === target) {
            return 0;
        }
        /**
         *    |
         * -------|-------
         */
        if (nums[mid] > nums[0]) {
            if (target > nums[mid]) {
                min = mid + 1;
            } else {
                if (target > nums[0]) {
                    max = mid - 1;
                } else {
                    min = mid + 1;
                }
            }
        }
        /**
         *            |
         * -------|-------
         */
        else {
            if (target < nums[mid]) {
                max = mid - 1;
            } else {
                if (target > nums[0]) {
                    max = mid - 1;
                } else {
                    min = mid + 1;
                }
            }
        }
    }

    return -1;
};
// @lc code=end

File: 33.搜索旋转排序数组.py
#
# @lc app=leetcode.cn id=33 lang=python3
#
# [33] 搜索旋转排序数组
#
from typing import List

# @lc code=start


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if not nums:
            return -1

        max = len(nums) - 1
        min = 0
        while max >= min:
            mid = (max + min) // 2
            if nums[mid] == target:
                return mid
            if max - min <= 1:
                return max if nums[max] == target else -1

            #   |
            # ---- | ----
            if nums[mid] >= nums[0]:
                if target >= nums[0] and target <= nums[mid]:
                    max = mid
                else:
                    min = mid
            #         |
            # ---- | ----
            if nums[mid] < nums[0]:
                if target < nums[0] and target >= nums[mid]:
                    min = mid
                else:
                    max = mid

        return -1


# @lc code=end
sol = Solution()
ans = sol.search([5, 1, 3], 5)
# ans = sol.search([4,5,6,7,0,1,2], 4)
print(ans)
